paddle/framework/lod_tensor.h

```diff
@@ -0,0 +1,226 @@
+  // then the smaller units level. Each level_t stores the offsets of units in
+  // Tesor.
+  typedef std::vector<level_t> lod_t;
+
+  LODTensor() {}
+  LODTensor(std::shared_ptr<Tensor> tensor, std::shared_ptr<lod_t> lod) {
+    Reset(tensor, lod);
+  }
+
+  void Reset(std::shared_ptr<Tensor> tensor, std::shared_ptr<lod_t> lod) {
+    tensor_ = tensor;
+    lod_start_pos_ = lod;
+  }
+
+  /*
+   * Get a element from LOD.
+   */
+  size_t lod_element(size_t level, size_t elem) const {
+    PADDLE_ENFORCE(level < Levels(), "level [%d] out of range [%d]", level,
+                   Levels());
+    PADDLE_ENFORCE(elem < Elements(level),
+                   "element begin [%d] out of range [%d]", elem,
+                   Elements(level));
+    return lod_start_pos_->at(level)[elem];
+  }
+
+  /*
+   * Number of LODTensor's levels, each level has units of data, for example,
+   * in the sentence's view, article, paragraph, sentence are 3 levels.
+   */
+  size_t Levels() const {
+    return lod_start_pos_ ? lod_start_pos_->size() : 0UL;
+  }
+  /*
+   * Number of elements in a level.
+   */
+  size_t Elements(size_t level = 0) const {
+    PADDLE_ENFORCE(level < Levels(), "level [%d] out of range [%d]", level,
+                   Levels());
+    // the last offset is the end of last element
+    return lod_start_pos_->at(level).size() - 1;
+  }
+
+  /*
+   * Slice of levels[level_begin:level_end], with tensor copied.
+   */
+  template <typename T>
+  LODTensor SliceCopied(size_t level_begin, size_t level_end,
+                        const platform::Place &dst_place) const;
+
+  /*
+   * Slice of levels[level_begin:level_end], with tensor shared.
+   */
+  LODTensor SliceShared(size_t level_begin, size_t level_end) const;
+
+  /*
+   * Slice of elements of a level, [elem_begin: elem_end], with tensor copied.
+   * @note: low performance in slice lod_start_pos_.
+   */
+  template <typename T>
+  LODTensor SliceCopied(size_t level, size_t elem_begin, size_t elem_end,
+                        const platform::Place &dst_place) const;
+
+  /*
+   * Slice of elements of a level, [elem_begin: elem_end], with tensor shared.
+   * @note: low performance in slice lod_start_pos_.
+   */
+  template <typename T>
+  LODTensor SliceShared(size_t level, size_t elem_begin, size_t elem_end) const;
+
+  /*
+   * Copy other's lod_start_pos_, to share LOD info.
+   * @note: the LOD info should not be changed.
+   */
+  void ShareLOD(const LODTensor &other) {
+    lod_start_pos_ = other.lod_start_pos_;
+  }
+
+  /*
+   * Copy other's lod_start_pos_'s content, free to mutate.
+   */
+  void CopyLOD(const LODTensor &other) {
+    lod_start_pos_ = std::make_shared<lod_t>(*other.lod_start_pos_);
+  }
+  /*
+   * Determine whether LODTensor has a valid LOD info.
+   */
+  bool has_lod() const { return lod_start_pos_.get(); }
+  std::shared_ptr<lod_t> const lod() const { return lod_start_pos_; }
```

Comment:
http://www.cplusplus.com/reference/memory/const_pointer_cast/

```cpp
std::shared_ptr<const lod_t> lod() const { return std::const_pointer_cast<const lod_t>(lod_start_pos_)}
```

However, is `get/set` methods are necessary?  The simplest code is

```cpp
public:
  std::shared_ptr<lod_t> lod_start_pos_;
  std::shared_ptr<Tensor> tensor_;
```